app/services/fact_cleaner.rb

```diff
@@ -39,11 +43,42 @@ def find_compose_orphaned_facts(found)
   end
 
   def delete!(to_delete)
-    logger.info "#{to_delete.size} facts will be removed"
+    logger.debug "#{to_delete.size} facts will be removed"
     to_delete.each do |fact|
       logger.debug { "Removing fact #{fact}" }
       fact.destroy
     end
-    logger.info "Cleanup of facts finished!"
+    logger.debug "Cleanup of facts finished!"
+  end
+
+  def find_excluded_facts
+    logger.debug "Searching for facts that match excluded pattern..."
+    fact_name_ids = []
+
+    excluded_facts.each do |exclude_pattern|
+      sql_pattern = exclude_pattern.gsub('*', '%')
+      # for facts that look like ignore_me::something
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', sql_pattern).pluck(:id)
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', sql_pattern + '::%').pluck(:id)
+      # for facts that look like common::ignore_me::something
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', '%::' + sql_pattern).pluck(:id)
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', '%::' + sql_pattern + '::%').pluck(:id)
+    end
+    fact_name_ids
+  end
+
+  def excluded_facts
+    Setting[:excluded_facts]
+  end
+
+  def delete_excluded_facts
+    fact_names = find_excluded_facts
+
+    # delete related fact values first
+    logger.debug "Removing values of excluded facts..."
+    FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
+    logger.debug "Removing excluded fact names..."
+    FactName.unscoped.where(:id => fact_names).delete_all
+    fact_names.count
```

Comment:
Seriously, we should not be doing regular cleaning for every request, but pull this to external process. This is way too many operations.

Reply:
Disregard this comment.